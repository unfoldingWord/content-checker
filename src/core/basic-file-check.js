import { isWhitespace, countOccurrences } from './text-handling-functions'

const CHECKER_VERSION_STRING = '0.0.1';

const DEFAULT_EXTRACT_LENGTH = 10;


export function doBasicFileChecks(filename, fileText, optionalFileLocation, optionalCheckingOptions) {
    // Does basic checks for small errors like mismatched punctuation pairs, etc.

    // filename (str): Used for identification
    // fileText (str): The field being checked
    // allowedLinks (bool): doesn't check links -- only checks lack of links
    // optionalFileLocation (str): Used to inform where this field is located

    // We assume that checking for compulsory fields is done elsewhere

    // Returns a single notice list
    //  The list contains arrays of five objects:
    //      1/ the priority number 0..999 (usually 800+ are errors, lower are warnings)
    //      2/ the error description string
    //      3/ the 0-based index for the position in the string (or -1 if irrelevant)
    //      4/ a short extract of the string containing the error (or empty-string if irrelevant)
    //      5/ the detailed location string
    //  (Returned in this way for more intelligent processing at a higher level)
    // console.log("doBasicFileChecks(" + filename + ", " + fileText.length.toLocaleString() + " chars, " +allowedLinks +", '"+ optionalFileLocation + "')…");
    console.assert(filename!==undefined, "doBasicFileChecks: 'filename' parameter should be defined");
    console.assert(typeof filename==='string', "doBasicFileChecks: 'filename' parameter should be a number not a '"+(typeof filename)+"': "+filename);
    console.assert(fileText!==undefined, "doBasicFileChecks: 'fileText' parameter should be defined");
    console.assert(typeof fileText==='string', "doBasicFileChecks: 'fileText' parameter should be a number not a '"+(typeof fileText)+"': "+fileText);
    // console.assert( allowedLinks===true || allowedLinks===false, "doBasicFileChecks: allowedLinks parameter must be either true or false");

    let result = { noticeList: [] };

    function addNotice(priority, message, index, extract, location) {
        // console.log("dBTC Notice: (priority="+priority+") "+message+(index > 0 ? " (at character " + index + 1 + ")" : "") + (extract ? " " + extract : "") + location);
        console.assert(priority!==undefined, "dBTCs addNotice: 'priority' parameter should be defined");
        console.assert(typeof priority==='number', "dBTCs addNotice: 'priority' parameter should be a number not a '"+(typeof priority)+"': "+priority);
        console.assert(message!==undefined, "dBTCs addNotice: 'message' parameter should be defined");
        console.assert(typeof message==='string', "dBTCs addNotice: 'message' parameter should be a string not a '"+(typeof message)+"': "+message);
        console.assert(index!==undefined, "dBTCs addNotice: 'index' parameter should be defined");
        console.assert(typeof index==='number', "dBTCs addNotice: 'index' parameter should be a number not a '"+(typeof index)+"': "+index);
        console.assert(extract!==undefined, "dBTCs addNotice: 'extract' parameter should be defined");
        console.assert(typeof extract==='string', "dBTCs addNotice: 'extract' parameter should be a string not a '"+(typeof extract)+"': "+extract);
        console.assert(location!==undefined, "dBTCs addNotice: 'location' parameter should be defined");
        console.assert(typeof location==='string', "dBTCs addNotice: 'location' parameter should be a string not a '"+(typeof location)+"': "+location);
        result.noticeList.push([priority, message, index, extract, location]);
    }


    // Main code for doBasicFileChecks()
    if (!fileText) // Nothing to check
        return result;

    // Create our more detailed location string by prepending the filename
    let ourAtString = " in '" + filename + "'";
    if (optionalFileLocation) {
        if (optionalFileLocation[0] != ' ') ourAtString += ' ';
        ourAtString += optionalFileLocation;
    }

    if (isWhitespace(fileText)) {
        addNotice(638, "Only found whitespace", -1, "", ourAtString);
        return result;
    }

    let extractLength;
    try {
        extractLength = optionalCheckingOptions.extractLength;
    } catch (e) {}
    if (typeof extractLength != 'number' || isNaN(extractLength)) {
        extractLength = DEFAULT_EXTRACT_LENGTH;
        // console.log("Using default extractLength=" + extractLength);
    }
    // else
    //     console.log("Using supplied extractLength=" + extractLength, "cf. default="+DEFAULT_EXTRACT_LENGTH);
    const halfLength = Math.floor(extractLength / 2); // rounded down
    const halfLengthPlus = Math.floor((extractLength+1) / 2); // rounded up
    // console.log("Using halfLength=" + halfLength, "halfLengthPlus="+halfLengthPlus);

    let ix = fileText.indexOf('<<<<<<<');
    if (ix >= 0) {
        let iy = ix+halfLength; // Want extract to focus more on what follows
        const extract = (iy>halfLength ? '…' : '') + fileText.substring(iy-halfLength, iy+halfLengthPlus).replace(/ /g, '␣') + (iy+halfLengthPlus < fileText.length ? '…' : '')
        addNotice(993, "Unresolved GIT conflict", ix, extract, ourAtString);
    } else {
        ix = fileText.indexOf('=======');
        if (ix >= 0) {
            let iy = ix+halfLength; // Want extract to focus more on what follows
            const extract = (iy>halfLength ? '…' : '') + fileText.substring(iy-halfLength, iy+halfLengthPlus).replace(/ /g, '␣') + (iy+halfLengthPlus < fileText.length ? '…' : '')
            addNotice(992, "Unresolved GIT conflict", ix, extract, ourAtString);
        } else {
            ix = fileText.indexOf('>>>>>>>>');
            if (ix >= 0) {
                let iy = ix+halfLength; // Want extract to focus more on what follows
                const extract = (iy>halfLength ? '…' : '') + fileText.substring(iy-halfLength, iy+halfLengthPlus).replace(/ /g, '␣') + (iy+halfLengthPlus < fileText.length ? '…' : '')
                addNotice(991, "Unresolved GIT conflict", ix, extract, ourAtString);
            }
        }
    }

    // Check matched pairs
    for (let punctSet of [['[',']'], ['(',')'], ['{','}'],
                          ['<','>'], ['⟨','⟩'], ['“','”'],
                          ['‹','›'], ['«','»'], ['**_','_**']]) {
        // Can't check '‘’' coz they might be used as apostrophe
        const leftChar = punctSet[0], rightChar = punctSet[1];
        const lCount = countOccurrences(fileText, leftChar);
        const rCount = countOccurrences(fileText, rightChar);
        if (lCount != rCount)
            addNotice(663, "Mismatched " + leftChar+rightChar + " characters", -1, "(left=" + lCount.toLocaleString() + ", right=" + rCount.toLocaleString() + ")", ourAtString);
    }

    // if (!allowedLinks) {
    //     // Simple check that there aren't any
    //     ix = fileText.indexOf('://');
    //     if (ix == -1) ix = fileText.indexOf('http');
    //     if (ix == -1) ix = fileText.indexOf('ftp');
    //     // The following might have to be removed if text fields can contain email addresses
    //     if (ix == -1) ix = fileText.indexOf('.org');
    //     if (ix == -1) ix = fileText.indexOf('.com');
    //     if (ix == -1) ix = fileText.indexOf('.info');
    //     if (ix == -1) ix = fileText.indexOf('.bible');
    //     if (ix >= 0) {
    //         let extract = (ix>halfLength ? '…' : '') + fileText.substring(ix-halfLength, ix+halfLengthPlus) + (ix+halfLengthPlus < fileText.length ? '…' : '')
    //         addNotice(765, "Unexpected link", ix, extract, ourAtString);
    //     }
    // }
    return result;
}
// end of doBasicFileChecks function

export default doBasicFileChecks;
