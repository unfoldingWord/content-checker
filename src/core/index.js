export * from './books/';
export * from './text-handling-functions';
export * from './basic-text-check';
export * from './markdown-text-check';
export * from './plain-text-check';
export * from './tn-table-row-check';
export * from './tn-table-text-check';
export * from './annotation-row-check';
export * from './annotation-table-check';
export * from './usfm-text-check';
export * from './notice-processing-functions';
export * from './book-package-check';
export * from './utilities';
