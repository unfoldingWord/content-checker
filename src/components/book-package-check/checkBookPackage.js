import * as books from '../../core/books/books';
import checkFile from '../file-check/checkFile';
import { consoleLogObject } from '../../core/utilities';
import { getFile } from '../../core/getApi';

const CHECKER_VERSION_STRING = '0.0.2';


async function checkBookPackage(username, language_code, bookCode, checkingOptions) {
    console.log("I'm here in checkBookPackage v" + CHECKER_VERSION_STRING);
    console.log("  with " + username + ", " + language_code + ", " + bookCode + ", " + JSON.stringify(checkingOptions));

    let checkBookPackageResult = { successList: [], noticeList: [] };

    function addSuccessMessage(successString) {
        console.log("Success: " + successString);
        checkBookPackageResult.successList.push(successString);
    }

    function addNotice(priority, message, index, extract, location, extra) {
        // console.log("checkBookPackage Notice: (priority=" + priority + ") " + message + (index > 0 ? " (at character " + index + 1 + ")" : "") + (extract ? " " + extract : "") + location);
        console.assert(typeof priority === 'number', "cBP addNotice: 'priority' parameter should be a number not a '" + (typeof priority) + "'");
        console.assert(priority !== undefined, "cBP addNotice: 'priority' parameter should be defined");
        console.assert(typeof message === 'string', "cBP addNotice: 'message' parameter should be a string not a '" + (typeof message) + "'");
        console.assert(message !== undefined, "cBP addNotice: 'message' parameter should be defined");
        console.assert(typeof index === 'number', "cBP addNotice: 'index' parameter should be a number not a '" + (typeof index) + "'");
        console.assert(index !== undefined, "cBP addNotice: 'index' parameter should be defined");
        console.assert(typeof extract === 'string', "cBP addNotice: 'extract' parameter should be a string not a '" + (typeof extract) + "'");
        console.assert(extract !== undefined, "cBP addNotice: 'extract' parameter should be defined");
        console.assert(typeof location === 'string', "cBP addNotice: 'location' parameter should be a string not a '" + (typeof location) + "'");
        console.assert(location !== undefined, "cBP addNotice: 'location' parameter should be defined");
        console.assert(typeof extra === 'string', "cBP addNotice: 'extra' parameter should be a string not a '" + (typeof extra) + "'");
        console.assert(extra !== undefined, "cBP addNotice: 'extra' parameter should be defined");
        checkBookPackageResult.noticeList.push([priority, message, index, extract, location, extra]);
    }

    function doOurCheckFile(repoCode, filename, file_content, fileLocation, optionalCheckingOptions) {
        // We assume that checking for compulsory fields is done elsewhere
        console.log("checkBookPackage doOurCheckFile(" + filename + ")");

        // Updates the global list of notices
        console.assert(typeof repoCode === 'string', "doOurCheckFile: 'repoCode' parameter should be a string not a '" + (typeof repoCode) + "'");
        console.assert(repoCode !== undefined, "doOurCheckFile: 'repoCode' parameter should be defined");
        console.assert(typeof filename === 'string', "doOurCheckFile: 'filename' parameter should be a string not a '" + (typeof filename) + "'");
        console.assert(filename !== undefined, "doOurCheckFile: 'filename' parameter should be defined");
        console.assert(typeof file_content === 'string', "doOurCheckFile: 'file_content' parameter should be a string not a '" + (typeof file_content) + "'");
        console.assert(file_content !== undefined, "doOurCheckFile: 'file_content' parameter should be defined");
        console.assert(typeof fileLocation === 'string', "doOurCheckFile: 'fileLocation' parameter should be a string not a '" + (typeof fileLocation) + "'");
        console.assert(fileLocation !== undefined, "doOurCheckFile: 'fileLocation' parameter should be defined");

        const resultObject = checkFile(filename, file_content, fileLocation, optionalCheckingOptions);
        console.log("checkFile() returned", resultObject.successList.length, "success message(s) and", resultObject.noticeList.length, "notice(s)");

        for (let successEntry of resultObject.successList)
            console.log("  ", successEntry);

        // Process results line by line,  appending the repoCode as an extra field as we go
        for (let noticeEntry of resultObject.noticeList)
            // noticeEntry is an array of five fields: 1=priority, 2=msg, 3=index, 4=extract, 5=location
            if (noticeEntry[0] != 663) // Suppress these misleading warnings coz open quote can occur in one verse and close in another
                // We add the repoCode as an extra value
                addNotice(noticeEntry[0], noticeEntry[1], noticeEntry[2], noticeEntry[3], noticeEntry[4], repoCode);
    }
    // end of doOurCheckFile function


    // Main code for checkBookPackage()
    const generalLocation = " " + language_code + " " + bookCode + " book package from " + username;
    let bookNumberAndName, whichTestament;
    try {
        bookNumberAndName = books.usfmNumberName(bookCode);
        whichTestament = books.testament(bookCode);
    } catch (e) {
        addNotice(900, "Bad function call: should be given a valid book abbreviation", -1, bookCode, " (not '" + bookCode + "')" + location);
        return checkBookPackageResult;
    }
    console.log("book_number_name='" + bookNumberAndName + "'", whichTestament);

    // No point in passing the branch through as a parameter
    //  coz if it's not 'master', it's unlikely to be common for all the repos
    const branch = 'master'

    // So now we want to work through checking this one specified Bible book in various repos:
    //  UHB/UGNT, ULT, UST, UTN, UTW, UTQ
    let checkedFileCount = 0, checkedFilenames = [], totalCheckedSize = 0, checkedRepoNames = [];
    for (let repoCode of [(whichTestament == 'old' ? 'UHB' : 'UGNT'), 'ULT', 'UST', 'TN', 'TQ']) {
        console.log("Let's try", repoCode, "(", language_code, bookCode, "from", username, ")");
        const repoLocation = " in " + repoCode.toUpperCase() + generalLocation;

        let repo_language_code = language_code;
        if (repoCode == 'UHB') repo_language_code = 'hbo';
        else if (repoCode == 'UGNT') repo_language_code = 'el-x-koine';
        const repoName = repo_language_code + '_' + repoCode.toLowerCase();

        const fullRepoName = username + '/' + repoName;
        console.log("Let's try1", bookCode, "from", fullRepoName);

        let filename;
        if (repoCode == 'UHB' || repoCode == 'UGNT' || repoCode == 'ULT' || repoCode == 'UST')
            filename = bookNumberAndName + '.usfm';
        else if (repoCode == 'TN')
            filename = language_code + '_tn_' + bookNumberAndName + '.tsv';
        else if (repoCode == 'TQ')
            // How are we going to handle all these folders of .md files ???
            // This resource will eventually be converted to TSV tables
            filename = bookCode.toLowerCase() + '/01/02.md';
        // console.log("Need to load", filename, "from", fullRepoName, generalLocation);

        console.log("Try to load", username, repoName, filename, branch);
        try {
            const fileContent = await getFile({ username, repository: repoName, path: filename, branch });
            console.log("Fetched file_content for", repoName, filename, typeof fileContent, fileContent.length);
            checkedFilenames.push(filename);
            totalCheckedSize += fileContent.length;
            checkedRepoNames.push(repoCode);
        } catch (e) {
            console.log("Failed to load", username, repoName, filename, branch, e + '');
            addNotice(996, "Failed to load", -1, "", generalLocation + " " + filename + ": " + e + '', repoCode);
            continue;
        }

        // We use the generalLocation here (does not include repo name)
        //  so that we can adjust the returned strings ourselves
        doOurCheckFile(repoCode, filename, fileContent, generalLocation, checkingOptions); // Adds the notices to checkBookPackageResult
        checkedFileCount += 1;
        addSuccessMessage(checkedFileCount + "/ Checked " + repoCode.toUpperCase() + " file: " + filename);
    }

    // Add some extra fields to our checkFileResult object
    //  in case we need this information again later
    checkBookPackageResult.checkedFileCount = checkedFileCount;
    checkBookPackageResult.checkedFilenames = checkedFilenames;
    checkBookPackageResult.checkedFilesizes = totalCheckedSize;
    checkBookPackageResult.checkedRepoNames = checkedRepoNames;
    // checkBookPackageResult.checkedOptions = checkingOptions; // This is done at the caller level

    console.log("checkBookPackage() is returning", checkBookPackageResult.successList.length.toLocaleString(), "success message(s) and", checkBookPackageResult.noticeList.length.toLocaleString(), "notice(s)");
    return checkBookPackageResult;
};
// end of checkBookPackage()

export default checkBookPackage;
