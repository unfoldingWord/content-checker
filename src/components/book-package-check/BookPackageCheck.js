import React, { useState, useEffect } from 'react';
import { withStyles } from '@material-ui/core/styles';
import * as books from '../../core/books/books';
import checkBookPackage from './checkBookPackage';
import processNoticesToErrorsWarnings from '../../core/notice-processing-functions';
import { RenderSuccessesErrorsWarnings } from '../RenderProcessedResults';
import { ourParseInt, consoleLogObject } from '../../core/utilities';


const CHECKER_VERSION_STRING = '0.1.1';


function BookPackageCheck(/*username, language_code, bookCode,*/ props) {
    // Check a single Bible book across many repositories
    const [result, setResultValue] = useState("Waiting-CheckBookPackage");

    // console.log(`I'm here in BookPackageCheck v${CHECKER_VERSION_STRING}`);
    // consoleLogObject("props", props);
    // consoleLogObject("props.classes", props.classes);

    let username = props.username;
    // console.log(`username='${username}'`);
    let language_code = props.language_code;
    // console.log(`language_code='${language_code}'`);
    let bookCode = props.bookCode;
    // console.log(`bookCode='${bookCode}'`);
    let branch = props.branch;
    // console.log(`branch='${branch}'`);

    if (bookCode!=='OBS' && !books.isValidBookCode(bookCode))
        return (<p>Please enter a valid USFM book code. ('{bookCode}' is not valid.)</p>);

     // Clear cached files if we've changed repo
    //  autoClearCache(bookCode); // This technique avoids the complications of needing a button

     let checkingOptions = { // Uncomment any of these to test them
        // 'extractLength': 25,
        };
    // Or this allows the parameters to be specified as a BookPackageCheck property
    if (props.extractLength) checkingOptions.extractLength = ourParseInt(props.extractLength);

    useEffect(() => {
        // Use an IIFE (Immediately Invoked Function Expression)
        //  e.g., see https://medium.com/javascript-in-plain-english/https-medium-com-javascript-in-plain-english-stop-feeling-iffy-about-using-an-iife-7b0292aba174
        (async () => {
            // console.log("Started unnamedFunction()");

            // Display our "waiting" message
            setResultValue(<p style={{ color: 'magenta' }}>Waiting for check results for {username} {language_code} <b>{bookCode}</b> book packageâ€¦</p>);

            const rawResult = await checkBookPackage(username, language_code, bookCode, setResultValue, checkingOptions);
            // console.log("checkBookPackage() returned", typeof rawResult); //, JSON.stringify(rawResult));

            // Add some extra fields to our rawResult object in case we need this information again later
            rawResult.checkType = 'BookPackage';
            rawResult.username = username;
            rawResult.language_code = language_code;
            rawResult.bookCode = bookCode;
            rawResult.checkedOptions = checkingOptions;

            // console.log("Here with CBP rawResult", typeof rawResult);
            // Now do our final handling of the result -- we have some options available
            let processOptions = { // Uncomment any of these to test them
                // 'maximumSimilarMessages': 3, // default is 2
                // 'errorPriorityLevel': 800, // default is 700
                // 'cutoffPriorityLevel': 100, // default is 0
                // 'sortBy': 'ByPriority', // default is 'AsFound'
                // 'ignorePriorityNumberList': [123, 202], // default is []
            };
            // Or this allows the parameters to be specified as a BookPackageCheck property
            if (props.maximumSimilarMessages) processOptions.maximumSimilarMessages = ourParseInt(props.maximumSimilarMessages);
            if (props.errorPriorityLevel) processOptions.errorPriorityLevel = ourParseInt(props.errorPriorityLevel);
            if (props.cutoffPriorityLevel) processOptions.cutoffPriorityLevel = ourParseInt(props.cutoffPriorityLevel);
            if (props.sortBy) processOptions.sortBy = props.sortBy;
            // if (props.ignorePriorityNumberList) processOptions.ignorePriorityNumberList = props.ignorePriorityNumberList;
            const processedResult = processNoticesToErrorsWarnings(rawResult, processOptions);
//             console.log(`BookPackageCheck got back processedResult with ${processedResult.successList.length.toLocaleString()} success message(s), ${processedResult.errorList.length.toLocaleString()} error(s) and ${processedResult.warningList.length.toLocaleString()} warning(s)
//   numIgnoredNotices=${processedResult.numIgnoredNotices.toLocaleString()} numSuppressedErrors=${processedResult.numSuppressedErrors.toLocaleString()} numSuppressedWarnings=${processedResult.numSuppressedWarnings.toLocaleString()}`);

            // console.log("Here now in rendering bit!");

            function renderSummary() {
                return (<>
                <p>Checked <b>{username} {language_code} {bookCode}</b> (from <i>{branch === undefined ? 'DEFAULT' : branch}</i> branches)</p>
                <p>&nbsp;&nbsp;&nbsp;&nbsp;Successfully checked {processedResult.checkedFileCount.toLocaleString()} file{processedResult.checkedFileCount==1?'':'s'} from {processedResult.checkedRepoNames.join(', ')}
                <br/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;including {processedResult.checkedFilenameExtensions.length} file type{processedResult.checkedFilenameExtensions.size === 1 ? '' : 's'}: {processedResult.checkedFilenameExtensions.join(', ')}.</p>
                </>);
            }

            if (processedResult.errorList.length || processedResult.warningList.length)
                setResultValue(<>
                    <p>{renderSummary()}
                        {processedResult.numIgnoredNotices ? ` (but ${processedResult.numIgnoredNotices.toLocaleString()} ignored errors/warnings)` : ""}</p>
                    <RenderSuccessesErrorsWarnings results={processedResult} />
                </>);
            else // no errors or warnings
                setResultValue(<>
                    <p>{renderSummary()}
                    {processedResult.numIgnoredNotices ? ` (with a total of ${processedResult.numIgnoredNotices.toLocaleString()} notices ignored)` : ""}</p>
                    <RenderSuccessesErrorsWarnings results={processedResult} />
                </>);

            // console.log("Finished rendering bit.");
        })(); // end of async part in unnamedFunction
    }, []); // end of useEffect part

    // {/* <div className={classes.root}> */}
    return (
        <div className="Fred">
        {result}
        </div>
    );
}

// BookPackageCheck.propTypes = {
//   /** @ignore */
//   username: PropTypes.object.isRequired,
//   /** @ignore */
//   language_code: PropTypes.object.isRequired,
//   bookCode: PropTypes.object.isRequired,
//   props: PropTypes.object,
// };

const styles = theme => ({
  root: {
  },
});

export default withStyles(styles)(BookPackageCheck);
