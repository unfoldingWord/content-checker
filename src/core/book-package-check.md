## Door43 Book Package Check Sandbox

This function checks the Door43 Book Package (i.e., a Bible book or Open Bible Stories) for the specified language by loading and checking files from several interconnected Door43 Content Service (DCS) repositories.

It returns a list of success messages and a list of notice components. (There is always a priority number in the range 0..999 and the main message string, as well as other details to help locate the error as available.)

These raw notice components can then be filtered and/or sorted as required by the calling program, and then divided into a list of errors and a list of warnings or whatever as desired.

The code below requests some info and then checks the single specified Bible book in several repos. This is convenient to see all these check results collected into one place.

See a list of valid book identifiers [here](http://ubsicap.github.io/usfm/identification/books.html), although only `GEN` to `REV` from that list are useful here.

Note that `OBS` can also be entered here as a *pseudo book identifier* in order to check an **Open Bible Stories** repo.

`Book Package Check` calls `checkBookPackage()` which then calls `checkFileContents()` for the book file in each repo (or calls `checkRepo()` for **OBS**).

**Warning**: Some book packages contain many files and/or very large files, and downloading them all and then checking them might slow down your browser -- maybe even causing pop-up messages asking to confirm that you want to keep waiting.

**Note**: This demonstration uses cached values of files stored inside the local browser. This makes reruns of the checks much faster, but it won't notice if you have updated the files on Door43. If you want to clear the local caches, use the `Clear Cache` function.

```js
import React, { useState, useEffect } from 'react';
import { checkBookPackage } from './book-package-check';
import { RenderRawResults } from '../demos/RenderProcessedResults';

// You can put your own data into the following fields:
const data = {
  username: 'unfoldingWord',
  languageCode : 'en',
  bookID : 'OBA',
  givenLocation : 'that was supplied',
  checkingOptions: {},
}

function CheckBookPackage(props) {
  const { username, languageCode, bookID, givenLocation, checkingOptions } = props.data;

  const [results, setResults] = useState(null);

  // We need the following construction because checkBookPackage is an ASYNC function
  useEffect(() => {
    // Use an IIFE (Immediately Invoked Function Expression)
    //  e.g., see https://medium.com/javascript-in-plain-english/https-medium-com-javascript-in-plain-english-stop-feeling-iffy-about-using-an-iife-7b0292aba174
    (async () => {
      // Display our "waiting" message
      setResults(<p style={{ color: 'magenta' }}>Checking <b>{username}</b> {languageCode} <b>{bookID}</b>â€¦</p>);
      const rawResults = await checkBookPackage(username, languageCode, bookID, setResults, checkingOptions);
      setResults(
        <div>
          <b>Checked</b> Door43 {username} {languageCode} {bookID}<br/><br/>
          <RenderRawResults results={rawResults} />
        </div>
      );
    })(); // end of async part in unnamedFunction
  }, []); // end of useEffect part

  return results;
} // end of CheckBookPackage function

<CheckBookPackage data={data}/>
```
