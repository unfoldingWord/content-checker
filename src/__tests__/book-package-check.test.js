/* eslint-env jest */

import { checkBookPackage } from '../core/book-package-check';
import Path from "path";
import fs from 'fs-extra';

let testFiles = {};

const optionalCheckingOptions = {
  getFile: params => {
    const { username, repository, path } = params;
    const filePath = Path.join('./src/__tests__/fixtures', username, repository, path);

    if (testFiles.hasOwnProperty(filePath)) { // see if we have a test file to use
      if (testFiles[filePath] !== null) {  // if file content not null, then return contents.  Otherwise will throw exception
        return testFiles[filePath];
      }
      throw `Simulated error - Could not find ${filePath}`;
    } else if (fs.existsSync(filePath)) {
      return fs.readFileSync(filePath).toString();
    }
    throw `Could not find ${filePath}`;
  }
}

describe('checkBookPackage()', () => {
  beforeEach(() => {
    testFiles = {}; // reset test files
  });

  it('tit should fail on unsupported language', async() => {
    const username = 'unfoldingWord';
    const languageCode = 'zzz';
    const bookID = 'tit';
    const rawResults = await checkBookPackage(username, languageCode, bookID, () => {}, optionalCheckingOptions);
    expect(rawResults.noticeList.length).toBeGreaterThan(0);
    const filteredResults = {
      successList: rawResults.successList,
      noticeList: rawResults.noticeList,
      checkedFilenames: rawResults.checkedFilenames,
      checkedRepoNames: rawResults.checkedRepoNames,
    };
    expect(filteredResults).toMatchSnapshot();
  });

  it('tit should fail on missing repo', async() => {
    const username = 'unfoldingWord';
    const languageCode = 'en';
    const bookID = 'tit';
    testFiles = { // override these files
      'src/__tests__/fixtures/unfoldingWord/en_ult/57-TIT.usfm': null,
    };

    const rawResults = await checkBookPackage(username, languageCode, bookID, () => {}, optionalCheckingOptions);
    expect(rawResults.noticeList.length).toBeGreaterThan(0);
    const filteredResults = {
      successList: rawResults.successList,
      noticeList: rawResults.noticeList,
      checkedFilenames: rawResults.checkedFilenames,
      checkedRepoNames: rawResults.checkedRepoNames,
    };
    expect(filteredResults).toMatchSnapshot();
  });

  it('tit should pass', async() => {
    const username = 'unfoldingWord';
    const languageCode = 'en';
    const bookID = 'tit';
    const rawResults = await checkBookPackage(username, languageCode, bookID, () => {}, optionalCheckingOptions);
    expect(rawResults.noticeList.length).toEqual(0);
    const filteredResults = {
      successList: rawResults.successList,
      noticeList: rawResults.noticeList,
      checkedFilenames: rawResults.checkedFilenames,
      checkedRepoNames: rawResults.checkedRepoNames,
    };
    expect(filteredResults).toMatchSnapshot();
  });

  it('rut should pass', async() => {
    const username = 'unfoldingWord';
    const languageCode = 'en';
    const bookID = 'rut';
    const rawResults = await checkBookPackage(username, languageCode, bookID, () => {}, optionalCheckingOptions);
    expect(rawResults.noticeList.length).toEqual(2);
    const filteredResults = {
      successList: rawResults.successList,
      noticeList: rawResults.noticeList,
      checkedFilenames: rawResults.checkedFilenames,
      checkedRepoNames: rawResults.checkedRepoNames,
    };
    expect(filteredResults).toMatchSnapshot();
  });

})
