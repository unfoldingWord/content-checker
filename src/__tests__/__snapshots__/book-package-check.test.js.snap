// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkBookPackage() -  rut should pass 1`] = `
Object {
  "checkedFilenames": Array [
    "08-RUT.usfm",
    "08-RUT.usfm",
    "08-RUT.usfm",
    "en_tn_08-RUT.tsv",
  ],
  "checkedRepoNames": Array [
    "UHB",
    "ULT",
    "UST",
    "TN",
  ],
  "noticeList": Array [
    Object {
      "C": "2",
      "V": "9",
      "bookID": "rut",
      "characterIndex": 52,
      "extra": "TN",
      "extract": "…ality - th…",
      "fieldName": "OccurrenceNote",
      "filename": "en_tn_08-RUT.tsv",
      "lineNumber": 92,
      "location": " in en rut book package from unfoldingWord master branch",
      "message": "Unexpected - character after space",
      "priority": 191,
      "rowID": "xc6u",
    },
    Object {
      "C": "4",
      "V": "14",
      "bookID": "rut",
      "characterIndex": 155,
      "extra": "TN",
      "extract": "…” or “ we …",
      "fieldName": "OccurrenceNote",
      "filename": "en_tn_08-RUT.tsv",
      "lineNumber": 255,
      "location": " in en rut book package from unfoldingWord master branch",
      "message": "Unexpected space after “ character",
      "priority": 192,
      "rowID": "ab47",
    },
  ],
  "successList": Array [
    "Checked UHB file: 08-RUT.usfm",
    "Checked ULT file: 08-RUT.usfm",
    "Checked UST file: 08-RUT.usfm",
    "Checked TN file: en_tn_08-RUT.tsv",
  ],
}
`;

exports[`checkBookPackage() -  tit should fail on missing repo 1`] = `
Object {
  "checkedFilenames": Array [
    "57-TIT.usfm",
    "57-TIT.usfm",
    "en_tn_57-TIT.tsv",
  ],
  "checkedRepoNames": Array [
    "UGNT",
    "UST",
    "TN",
  ],
  "noticeList": Array [
    Object {
      "bookID": "tit",
      "extra": "ULT",
      "filename": "57-TIT.usfm",
      "location": " in ULT in en tit book package from unfoldingWord master branch: Simulated error - Could not find src/__tests__/fixtures/unfoldingWord/en_ult/57-TIT.usfm",
      "message": "Failed to load",
      "priority": 996,
      "repoName": "en_ult",
    },
  ],
  "successList": Array [
    "Checked UGNT file: 57-TIT.usfm",
    "Checked UST file: 57-TIT.usfm",
    "Checked TN file: en_tn_57-TIT.tsv",
  ],
}
`;

exports[`checkBookPackage() -  tit should fail on unsupported language 1`] = `
Object {
  "checkedFilenames": Array [
    "57-TIT.usfm",
  ],
  "checkedRepoNames": Array [
    "UGNT",
  ],
  "noticeList": Array [
    Object {
      "bookID": "tit",
      "extra": "GLT",
      "filename": "57-TIT.usfm",
      "location": " in GLT in zzz tit book package from unfoldingWord master branch: Could not find src/__tests__/fixtures/unfoldingWord/zzz_glt/57-TIT.usfm",
      "message": "Failed to load",
      "priority": 996,
      "repoName": "zzz_glt",
    },
    Object {
      "bookID": "tit",
      "extra": "GST",
      "filename": "57-TIT.usfm",
      "location": " in GST in zzz tit book package from unfoldingWord master branch: Could not find src/__tests__/fixtures/unfoldingWord/zzz_gst/57-TIT.usfm",
      "message": "Failed to load",
      "priority": 996,
      "repoName": "zzz_gst",
    },
    Object {
      "bookID": "tit",
      "extra": "TN",
      "filename": "zzz_tn_57-TIT.tsv",
      "location": " in TN in zzz tit book package from unfoldingWord master branch: Could not find src/__tests__/fixtures/unfoldingWord/zzz_tn/zzz_tn_57-TIT.tsv",
      "message": "Failed to load",
      "priority": 996,
      "repoName": "zzz_tn",
    },
  ],
  "successList": Array [
    "Checked UGNT file: 57-TIT.usfm",
  ],
}
`;

exports[`checkBookPackage() -  tit should pass 1`] = `
Object {
  "checkedFilenames": Array [
    "57-TIT.usfm",
    "57-TIT.usfm",
    "57-TIT.usfm",
    "en_tn_57-TIT.tsv",
  ],
  "checkedRepoNames": Array [
    "UGNT",
    "ULT",
    "UST",
    "TN",
  ],
  "noticeList": Array [],
  "successList": Array [
    "Checked UGNT file: 57-TIT.usfm",
    "Checked ULT file: 57-TIT.usfm",
    "Checked UST file: 57-TIT.usfm",
    "Checked TN file: en_tn_57-TIT.tsv",
  ],
}
`;
